lucet-runtime/lucet-runtime-internals/src/instance.rs

```diff
@@ -531,24 +573,38 @@ impl Instance {
     // so that it is 8 bytes before the heap.
     // For this reason, the alignment of the structure is set to 4096, and we define accessors that
     // read/write the globals pointer as bytes [4096-8..4096] of that structure represented as raw bytes.
+    // InstanceRuntimeData is placed such that it ends at the end of the page this `Instance` starts
+    // on. So we can access it by *self + PAGE_SIZE - size_of::<InstanceRuntimeData>
     #[inline]
-    pub fn get_globals_ptr(&self) -> *const i64 {
+    fn get_instance_implicits(&self) -> &InstanceRuntimeData {
         unsafe {
-            *((self as *const _ as *const u8)
-                .offset((HOST_PAGE_SIZE_EXPECTED - mem::size_of::<*mut i64>()) as isize)
-                as *const *const i64)
+            let implicits_ptr = (self as *const _ as *const u8)
+                .offset((HOST_PAGE_SIZE_EXPECTED - mem::size_of::<InstanceRuntimeData>()) as isize)
+                as *const InstanceRuntimeData;
+            mem::transmute::<*const InstanceRuntimeData, &InstanceRuntimeData>(implicits_ptr)
         }
     }
 
     #[inline]
-    pub fn set_globals_ptr(&mut self, globals_ptr: *const i64) {
+    fn get_instance_implicits_mut(&mut self) -> &mut InstanceRuntimeData {
         unsafe {
-            *((self as *mut _ as *mut u8)
-                .offset((HOST_PAGE_SIZE_EXPECTED - mem::size_of::<*mut i64>()) as isize)
-                as *mut *const i64) = globals_ptr;
+            let implicits_ptr = (self as *mut _ as *mut u8)
+                .offset((HOST_PAGE_SIZE_EXPECTED - mem::size_of::<InstanceRuntimeData>()) as isize)
+                as *mut InstanceRuntimeData;
+            mem::transmute::<*mut InstanceRuntimeData, &mut InstanceRuntimeData>(implicits_ptr)
         }
     }
 
+    #[inline]
+    pub fn get_globals_ptr(&self) -> *mut i64 {
```

Comment:
These are `pub fn` in the `// Private API` part of `impl Instance { }` - are they lost stragglers or were they there intentionally? @acfoltzer ?

Reply:
Ah, I actually missed this when we made the change away from explicit padding. These should be private.